app/controllers/provider/admin/account/enforce_sso_controller.rb

```diff
@@ -7,17 +7,17 @@ class Provider::Admin::Account::EnforceSSOController < Provider::Admin::Account:
   def create
     if current_account.settings.update(enforce_sso: true)
       logout_all_password_users
-      redirect_to index_path, notice: 'SSO successfully enforced'
+      render json: { redirect: provider_login_path, notice: t('.success') }
```

Comment:
I changed it. From now on it should only redirect if current session uses password

Reply:
Why did you add this redirect? it seems to me no redirect is needed, since it's an Async request. Anyway, this code is supposed to be executed only when the user has been logged in through SSO, so `current_user.using_password?` should always be false. And if it were true, then that would be a bug. In my opinion, this should be just:

```ruby
render json: { notice: t('.success') }
```

Reply:
Actually, `if current_user.using_password?` will return `true` even if currently the user is logged in through SSO (in fact, otherwise the switch would not be enabled), it just indicates whether the user has the password stored in the DB.

The condition would have to be changed to `unless user_session.sso_login?`, I think. But this is redundant, because this same check is already performed by `enforce_sso_allowed?`. So I agree with @jlledom , let's just drop it.

Reply:
> it just indicates whether the user has the password stored in the DB.

But Oauth users don't have a password sotred in DB, or I am wrong?

Reply:
> But Oauth users don't have a password sotred in DB, or I am wrong?

There is no such thing as OAuth-only user. A user signed-up (or logged-in) through OAuth can change their password, and - if user-based password is not disabled - can log in later using that password.

And the other way around - a user may sign up using username-password, but then also log in using SSO (for example, if it is enabled later), both ways will be associated to the same user, by email.

Reply:
That screen looks super nice, I don't see a problem on it...

Reply:
@mayorova @jlledom 
For the record: I did not add any redirect. The redirect was there already, I just changed the path. To ensure backwards compatibility the redirect should stay..

Sessions "password only" will be destroyed, therefore there has to be a redirect because the current session could be done. It can also be a simple page refresh.

However, if the current or other session is "SSO" it is still valid and the page should not change.

Reply:
@josemigallas I must admit I am not sure how it was before, so I am not saying that this is something added by this PR. But my point is: when we disable password-based logins, the user is already logged in via SSO, so the redirect shouldn't happen, and last time I checked it was redirecting.

Reply:
It should only redirect if current session is password-based, but it should redirect. The problem without a redirect is that the user will think the session is still open until they click anywhere and then they will be redirected by rails.

You said earlier there's a method `user_session.sso_login?` to check this. It is different from `enforce_sso_allowed?`, not redundant. The latter checks whether the setting is set, not that the current session was opened via SSO. I will update the code.

Reply:
@josemigallas Well, `user_session.sso_login?` is **part** of `enforce_sso_allowed?` https://github.com/3scale/porta/blob/edce496fab6003bdee8ca0dfac37ccb799854332/app/validators/enforce_sso_validator.rb#L31-L33

(it DOES check that the current session is SSO-based, but it also checks other stuff - like the integration was tested recently etc.)

> It should only redirect if current session is password-based, but it should redirect.

Yes, but again - if the current session is password-based, the switch will be disabled :)  But the user of course can: 
- be logged in through password
- go to the HTML of the page and remove the `disabled` attr
- switch the switch

So yeah, we can keep that redirect for those cases! :) I was just saying that this would not be a standard flow.

Reply:
The current session must be sso-based, OK... I missed that part, so the redirect was silly and yet nobody ever noticed or got annoyed by it 😅